lib/matplotlib/_constrained_layout.py

```diff
@@ -83,181 +74,350 @@ def do_constrained_layout(fig, renderer, h_pad, w_pad,
-    for nn, hss in enumerate(hassubplotspec):
-        if not hss:
-            # this gridspec slot doesn't have an axis so we
-            # make a "ghost".
-            ax = fig.add_subplot(gs[nn])
-            ax.set_visible(False)
-
-
-def _make_layout_margins(ax, renderer, h_pad, w_pad):
+        if hasattr(ax, 'get_subplotspec'):
+            gs = ax.get_subplotspec().get_gridspec()
+            lg = gs._layoutgrid
+            if lg is not None:
+                for i in range(gs.nrows):
+                    for j in range(gs.ncols):
+                        bb = lg.get_inner_bbox(i, j)
+                        if bb.width <= 0 or bb.height <= 0:
+                            return False
+    return True
+
+
+def _get_margin_from_padding(object, *, w_pad=0, h_pad=0,
+                             hspace=0, wspace=0):
+
+    ss = object._subplotspec
+    gs = ss.get_gridspec()
+    lg = gs._layoutgrid
+
+    if hasattr(gs, 'hspace'):
+        _hspace = (gs.hspace if gs.hspace is not None else hspace)
+        _wspace = (gs.wspace if gs.wspace is not None else wspace)
+    else:
+        _hspace = (gs._hspace if gs._hspace is not None else hspace)
+        _wspace = (gs._wspace if gs._wspace is not None else wspace)
+
+    _wspace = _wspace / 2
+    _hspace = _hspace / 2
+
+    nrows, ncols = gs.get_geometry()
+    # there are two margins for each direction.  The "cb"
+    # margins are for pads and colorbars, the non-"cb" are
+    # for the axes decorations (labels etc).
+    margin = {'leftcb': w_pad, 'rightcb': w_pad,
+              'bottomcb': h_pad, 'topcb': h_pad,
+              'left': 0, 'right': 0,
+              'top': 0, 'bottom': 0}
+    if _wspace / ncols > w_pad:
+        if ss.colspan.start > 0:
+            margin['leftcb'] = _wspace / ncols
+        if ss.colspan.stop < ncols:
+            margin['rightcb'] = _wspace / ncols
+    if _hspace / nrows > h_pad:
+        if ss.rowspan.stop < nrows:
+            margin['bottomcb'] = _hspace / nrows
+        if ss.rowspan.start > 0:
+            margin['topcb'] = _hspace / nrows
+
+    return margin
+
+
+def _make_layout_margins(fig, renderer, *, w_pad=0, h_pad=0,
+                         hspace=0, wspace=0):
     """
     For each axes, make a margin between the *pos* layoutbox and the
     *axes* layoutbox be a minimum size that can accommodate the
     decorations on the axis.
+
+    Then make room for colorbars.
+    """
+    for panel in fig.panels:  # recursively make child panel margins
+        ss = panel._subplotspec
+        _make_layout_margins(panel, renderer, w_pad=w_pad, h_pad=h_pad,
+                             hspace=hspace, wspace=wspace)
+
+        margins = _get_margin_from_padding(panel, w_pad=0, h_pad=0,
+                                           hspace=hspace, wspace=wspace)
+        panel._layoutgrid.parent.edit_outer_margin_mins(margins, ss)
+
+    # for ax in [a for a in fig._localaxes if hasattr(a, 'get_subplotspec')]:
```

Comment:
remove?

Reply:
Well, I'm hoping to get a version in that uses `_localaxes` for 3.4, so this was just bread crumbing....